agent/src/beerocks/slave/son_slave_thread.cpp

```diff
@@ -3728,92 +3728,83 @@ bool slave_thread::slave_fsm(bool &call_slave_select)
-            notification->backhaul_params().bridge_ipv4 =
+            vs_tlv->backhaul_params().bridge_ipv4 =
                 network_utils::ipv4_from_string(backhaul_params.bridge_ipv4);
-            notification->backhaul_params().backhaul_ipv4 =
+            vs_tlv->backhaul_params().backhaul_ipv4 =
                 network_utils::ipv4_from_string(backhaul_params.bridge_ipv4);
         } else {
-            notification->backhaul_params().backhaul_ipv4 =
+            vs_tlv->backhaul_params().backhaul_ipv4 =
                 network_utils::ipv4_from_string(backhaul_params.backhaul_ipv4);
         }
 
         std::copy_n(backhaul_params.backhaul_scan_measurement_list,
                     beerocks::message::BACKHAUL_SCAN_MEASUREMENT_MAX_LENGTH,
-                    notification->backhaul_params().backhaul_scan_measurement_list);
+                    vs_tlv->backhaul_params().backhaul_scan_measurement_list);
 
         for (unsigned int i = 0; i < message::BACKHAUL_SCAN_MEASUREMENT_MAX_LENGTH; i++) {
-            if (notification->backhaul_params().backhaul_scan_measurement_list[i].channel > 0) {
+            if (vs_tlv->backhaul_params().backhaul_scan_measurement_list[i].channel > 0) {
                 LOG(DEBUG)
                     << "mac = "
-                    << network_utils::mac_to_string(notification->backhaul_params()
-                                                        .backhaul_scan_measurement_list[i]
-                                                        .mac.oct)
+                    << network_utils::mac_to_string(
+                           vs_tlv->backhaul_params().backhaul_scan_measurement_list[i].mac.oct)
                     << " channel = "
-                    << int(notification->backhaul_params()
-                               .backhaul_scan_measurement_list[i]
-                               .channel)
+                    << int(vs_tlv->backhaul_params().backhaul_scan_measurement_list[i].channel)
                     << " rssi = "
-                    << int(notification->backhaul_params().backhaul_scan_measurement_list[i].rssi);
+                    << int(vs_tlv->backhaul_params().backhaul_scan_measurement_list[i].rssi);
             }
         }
 
         //Platform Settings
-        notification->platform_settings() = platform_settings;
+        vs_tlv->platform_settings() = platform_settings;
 
         //Wlan Settings
-        notification->wlan_settings() = wlan_settings;
+        vs_tlv->wlan_settings() = wlan_settings;
         // Hostap Params
-        notification->hostap()          = hostap_params;
-        notification->hostap().ant_gain = config.hostap_ant_gain;
+        vs_tlv->hostap()          = hostap_params;
+        vs_tlv->hostap().ant_gain = config.hostap_ant_gain;
 
         // Channel Selection Params
-        notification->cs_params() = hostap_cs_params;
+        vs_tlv->cs_params() = hostap_cs_params;
 
-        vs->length() += notification->getLen();
```

Comment:
So this is why the first commit in the series is needed: it used to be done correctly, but here you break it...

I don't really like doing that hack in the finalize function. I agree that it's not so nice to have to do it explicitly everywhere we use the vs TLV, but I don't think the approach used here is that much better.

Perhaps we can add a wrapper class that does this update of length in its destructor? Something like this:

```cpp
template<class T> class beerocks_tlv {
    shared_ptr<T> action;
    shared_ptr<tlvVendorSpecific> vsTlv;
public:
    T& operator->() { return *action; }
    ~beerocks_tlv() {vsTlv->length() += action->getLen(); }
};
```

Reply:
I should have done the remove of the `vs->length() += notification->getLen();` in the first commit.
I don't quite understand the solution you are proposing, but if you think you can do a better solution, please create a PR about it and discuss about it.
For now, I will remove the update of the length in the first commit, which is the right place to do it.

Reply:
I probably won't have time to do this refactoring any time soon

However, what do you think of this pattern, of doing magic stuff in the destructor? The idea is to have a lightweight object that stays alive as long as you're editing the VS TLV, and when you're finished with it, the lightweight object goes out of scope and the TLV length gets automagically updated. It is still a bit magical so maybe not that ideal. And maybe an explicit finalize is better than relying on the destructor...